### DISCUSSION TITLE
.NET 8 support?

### DISCUSSION THREAD

**Discussion by bogdan-patraucean** [microsoft/WindowsAppSDK#3967] [Q&A] [answered]
How to update to .NET 8? Doesn't seem to work properly.
There is also an [issue ](https://github.com/microsoft/WindowsAppSDK/issues/3842#issuecomment-1811116958) with the RID graph.

**Comment by DarranRowe** [comment]
You should always be more specific with what you mean by "Doesn't seem to work." This could mean anything.
Assuming that you are using Visual Studio 2022 17.8.0 (this seems to help with the project losing the target framework identifier), then there are a few steps you need to follow.
1) Add UseRidGraph to the .csproj file. This step is optional, but it removes the constant NETSDK1206 warning.
2) Modify the runtime identifiers to remove the 10. So after both of these modifications, your .csproj file could look similar to:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>CsMeh</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;ARM64</Platforms>
    <!--Modified runtime identifiers-->
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <UseWinUI>true</UseWinUI>
  </PropertyGroup>

  <ItemGroup>
    <!--Added to put .NET 8 into backwards compatibility mode.-->
    <RuntimeHostConfigurationOption Include="System.Runtime.Loader.UseRidGraph" Value="true" />
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.230913002" />
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.755" />
    <Manifest Include="$(ApplicationManifest)" />
  </ItemGroup>
</Project>
```
3) Rename all of the publish profiles to match the respective runtime identifiers. For example, rename win10-x86.pubxml to win-x86.pubxml.
4) In each of the .pubxml files, modify the RuntimeIdentifier to remove the 10, in the same way as you did for the .csproj file.

Until there is a proper fix for this in the Windows App SDK itself, these are the steps you need to do to get the projects to build.

**Reply by bogdan-patraucean** [reply]
@DarranRowe sorry for the late reply, the app builds, but when to open, it doesn't. It hangs at the step before the debugging process starts. Also why do we need to specify the framework like this: 
`<TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>` 
instead of this:
 `<TargetFramework>net8.0</TargetFramework>`?
 
 I'm using the latest `VS 17.8`, yes, with latest `WinAppSdk 1.4.3`, packaged.

**Reply by DarranRowe** [reply]
>Also why do we need to specify the framework like this:
`<TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>`
instead of this:
`<TargetFramework>net8.0</TargetFramework>`?

[To access the C# interop classes](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/winrt-com-interop-csharp#in-other-c-desktop-project-types-wpf-or-winforms). There is a visible difference.

.net8.0:

![Screenshot 2023-11-18 215101](https://github.com/microsoft/WindowsAppSDK/assets/52577874/ec9ca169-9ebf-4509-af04-8721c023cc60)

.net8.0-windows:

![Screenshot 2023-11-18 215012](https://github.com/microsoft/WindowsAppSDK/assets/52577874/4938362e-e0e7-4794-ba6c-d3c4631768f2)

.net8.0-windows10.0.19041.0

![Screenshot 2023-11-18 215035](https://github.com/microsoft/WindowsAppSDK/assets/52577874/777186be-4f72-4271-ab36-41582feda474)

Now, this is just a C# console application which didn't reference the Windows App SDK at the time. But the big important thing that it enables is Windows Runtime (WinRT) interop. This is important for the Windows App SDK because it is backed by WinRT objects.

>sorry for the late reply, the app builds, but when to open, it doesn't. It hangs at the step before the debugging process starts.

I haven't experienced that. Do you get any debugger output at all? Do you get anything useful if you enable native debugging?

**Reply by bogdan-patraucean** [reply]
@DarranRowe I get nothing. Just hangs there. Might just be a Windows beta issue, I have no idea yet.

**Comment by Dev-Taehui** [comment]
It doesn't work when you set unpackaged mode. `Microsoft.WindowsAppRuntime.Bootstrap.dll` is missing after build. (NOT `Microsoft.WindowsAppRuntime.Bootstrap.NET.dll`) So, it crashed with DllNotFoundException

It works on packaged mode.

**Reply by DarranRowe** [reply]
There is certainly an interesting problem here.

![Screenshot 2023-11-21 021558](https://github.com/microsoft/WindowsAppSDK/assets/52577874/698b2ced-8c17-4fa7-804a-b56456d07c67)

If you use the WinUI 3 template with a separate packaging project, it certainly copies the bootstrapper into the application's directory. However, if you do not have UseRidGraph set in one form or another, then it is unable to find the runtimes in the runtimes subdirectory.

If you use the WinUI 3 template which does the packaging in place, then yes, it doesn't copy the bootstrapper. I would guess that it is trying to copy the file, but if you have a look in the Windows App SDK nuget package, the bootstrapper is in a path that contains win10. Out of curiosity, I did try renaming the directory in the nuget package to remove the 10, but that didn't work.

**Comment by lhak** [comment]
Adding UseRidGraph=true to the project file (without changing the runtime identifiers) seems to be enough to get it working (single project c#) in packaged and unpackaged configuration for me.

**Comment by bogdan-patraucean** [comment, marked as answer]
Support was [added](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#suppport-for-net-8).
